### PR TITLE
Fix last confirmed blocks

### LINKED ISSUE #272: darwinia-ethereum-relay ropsten/mainet network type
ropsten and mainet have differences in ethash params

### LINKED ISSUE #273: verified's receipt should not limit the mmr_root block number to last confirmed
https://github.com/darwinia-network/darwinia-common/blob/master/frame/bridge/ethereum/relay/src/lib.rs#L628

Currently,  ```MMRProof``` is
```
pub type MMRProof = Vec<H256>;
```

should be refactored to something like 

```
pub struct MMRProof {
   // member leaf
   u64 member_block_number,
   // the number of the block attaching mmr root
   u64 mmr_root_block_number,
   // mmr proof generated
   Vec<H256>
}
```

### LINKED ISSUE #275: Remove ConfirmedHeadersDoubleMap and ConfirmBlocksInCycle etc.
Using simple ConfirmedBlocks and MaxConfirmedBlocksCount instead.

### LINKED ISSUE #276: ethereum-relay's should have genesis header or should able to set last confirmed block using root origin

### PR DESCRIPTION
**Pull request by HackFisher**
- [x] Closes #275 
- [x] Closes #273 
- [x] Closes #276
- [x] Closes #272 
- [x] Fix Tests
- [x] Migration

### CODE REVIEW

File: frame/bridge/ethereum/relay/src/lib.rs [hunk 1]
```diff
@@ -693,6 +662,22 @@ impl HeaderThing for EthereumHeaderThing {
...
+	pub mmr_root: H256,
+}
+
+#[cfg_attr(any(feature = "deserialize", test), derive(serde::Deserialize))]
+#[derive(Clone, Default, PartialEq, Encode, Decode, RuntimeDebug)]
+pub struct MMRProof {
+	pub member_leaf_index: u64,
+	pub last_leaf_index: u64,
+	pub proof: Vec<H256>
+}
```
Comment (clearloop) [inline frame/bridge/ethereum/relay/src/lib.rs:679]: Should update this new struct in `types.json` together

Comment (HackFisher) [reply frame/bridge/ethereum/relay/src/lib.rs:752]: Updated

**Comment by AurevoirXavier** [comment]
I can't find anywhere read/write `ConfirmedDepth`.

**Comment by AurevoirXavier** [comment]
I found some read/write first check last, fixed.

Read last avoid the unnecessary DB reads.
Write last guarantee no side effect on storage.

File: frame/bridge/ethereum/relay/src/lib.rs [hunk 2]
```diff
@@ -112,6 +155,9 @@ decl_event! {
 
 		/// EthereumReceipt Verification. [account, receipt, header]
 		VerifyReceipt(AccountId, EthereumReceipt, EthereumHeader),
+
+		/// Reset Last Confirmed Header
+		ResetLastConfirmedHeader(EthereumHeaderThing),
```
Comment (AurevoirXavier) [inline frame/bridge/ethereum/relay/src/lib.rs:160]: Many unused events.

File: frame/bridge/ethereum/relay/src/lib.rs [hunk 3]
```diff
@@ -131,8 +177,8 @@ decl_error! {
 		MMRI,
 		/// Header Hash - MISMATCHED
 		HeaderHashMis,
-		/// Last Header - NOT EXISTED
-		LastHeaderNE,
+		/// Confirmed Header - NOT EXISTED
+		ConfirmedHeaderNE,
```
Comment (AurevoirXavier) [inline frame/bridge/ethereum/relay/src/lib.rs:181]: Many unused errors.

File: frame/bridge/ethereum/relay/src/lib.rs [hunk 4]
```diff
@@ -276,47 +326,46 @@ decl_module! {
...
-				// read the doc string of of event
-				Self::deposit_event(RawEvent::ConfirmBlockManagementError(month));
-			} else {
-				ConfirmBlocksInCycle::put(blocks_in_month);
-				ConfirmBlockKeepInMonth::put(month);
+			ConfirmedBlockNumbers::put(new_confirmed_blocks);
 
-				Self::deposit_event(RawEvent::UpdateConfrimedBlockCleanCycle(month, blocks_in_month));
-			}
+			Self::store_header(header_thing.clone())?;
```
Comment (AurevoirXavier) [inline frame/bridge/ethereum/relay/src/lib.rs:366]: Write first check last found.

File: frame/bridge/ethereum/relay/src/lib.rs [hunk 5]
```diff
@@ -330,37 +379,34 @@ impl<T: Trait> Module<T> {
...
-				(header.number.saturating_sub(1)) / confirm_blocks_in_cycle,
-				header.number.saturating_sub(1),
-			);
-			if header.parent_hash != previous_header.hash.unwrap_or_default()
-				|| *header.difficulty()
-					!= eth_partial.calculate_difficulty(header, &previous_header)
-			{
+	fn verify_block_with_last_confirmed_block(header: &EthereumHeader) -> bool {
+		let eth_partial = Self::ethash_params();
+		let last_confirmed_block = Self::best_block_number();
```
Comment (AurevoirXavier) [inline frame/bridge/ethereum/relay/src/lib.rs:392]: Read first check last found.

**Comment by HackFisher** [comment]
> I can't find anywhere read/write `ConfirmedDepth`.

ConfirmedDepth is in todo, will be implemented in later pr

**Review by AurevoirXavier — APPROVED** [review]

**Pull request merged**